api/src/test/java/org/apache/iceberg/TestPartitionSpecValidation.java

```diff
@@ -121,6 +123,66 @@ public void testMultipleDatePartitionsWithDifferentSourceColumns() {
+        .build().fields().get(0).name(), "timestamp");
+    Assert.assertEquals(PartitionSpec.builderFor(SCHEMA).year("ts", "custom_year")
+        .build().fields().get(0).name(), "custom_year");
+    Assert.assertEquals(PartitionSpec.builderFor(SCHEMA).month("ts", "custom_month")
+        .build().fields().get(0).name(), "custom_month");
+    Assert.assertEquals(PartitionSpec.builderFor(SCHEMA).day("ts", "custom_day")
+        .build().fields().get(0).name(), "custom_day");
+    Assert.assertEquals(PartitionSpec.builderFor(SCHEMA).hour("ts", "custom_hour")
+        .build().fields().get(0).name(), "custom_hour");
+    Assert.assertEquals(PartitionSpec.builderFor(SCHEMA)
+        .bucket("ts", 4, "custom_bucket")
+        .build().fields().get(0).name(), "custom_bucket");
+    Assert.assertEquals(PartitionSpec.builderFor(SCHEMA)
+        .truncate("s", 1, "custom_truncate")
+        .build().fields().get(0).name(), "custom_truncate");
+  }
+
+  @Test
+  public void testSettingPartitionTransformsWithCustomTargetNamesThatAlreadyExist() {
+
+    AssertHelpers.assertThrows("Should not allow target column name that exists in schema",
+        IllegalArgumentException.class,
+        "Cannot create partition from name that exists in schema: another_ts",
+        () -> PartitionSpec.builderFor(SCHEMA).year("ts", "another_ts"));
+
+    AssertHelpers.assertThrows("Should not allow target column name that exists in schema",
+        IllegalArgumentException.class,
+        "Cannot create partition from name that exists in schema: another_ts",
+        () -> PartitionSpec.builderFor(SCHEMA).month("ts", "another_ts"));
+
+    AssertHelpers.assertThrows("Should not allow target column name that exists in schema",
+        IllegalArgumentException.class,
+        "Cannot create partition from name that exists in schema: another_ts",
+        () -> PartitionSpec.builderFor(SCHEMA).day("ts", "another_ts"));
+
+    AssertHelpers.assertThrows("Should not allow target column name that exists in schema",
+        IllegalArgumentException.class,
+        "Cannot create partition from name that exists in schema: another_ts",
+        () -> PartitionSpec.builderFor(SCHEMA).hour("ts", "another_ts"));
+
+    AssertHelpers.assertThrows("Should not allow target column name that exists in schema",
+        IllegalArgumentException.class,
+        "Cannot create partition from name that exists in schema: another_ts",
+        () -> PartitionSpec.builderFor(SCHEMA).truncate("ts", 2, "another_ts"));
+
+    AssertHelpers.assertThrows("Should not allow target column name that exists in schema",
+        IllegalArgumentException.class,
+        "Cannot create partition from name that exists in schema: another_ts",
+        () -> PartitionSpec.builderFor(SCHEMA).bucket("ts", 4, "another_ts"));
+
+    // allows for identity
+    Assert.assertEquals(PartitionSpec.builderFor(SCHEMA).identity("ts", "another_ts")
+            .build().fields().get(0).name(),
+        "another_ts");
```

Comment:
Actually, I don't think this should be allowed either. This should only be allowed if the partition column is named after its own source column.